fido/Crypto/Fido2/Model/JavaScript/Encoding.hs

```diff
@@ -169,6 +179,80 @@ instance Encode (M.PublicKeyCredentialOptions 'M.Get) where
         extensions = encode pkcogExtensions
       }
 
+-- | [(spec)](https://www.w3.org/TR/webauthn-2/#iface-pkcredential)
+-- Encodes the PublicKeyCredential for attestation, this instance is mostly used in the tests where we emulate the
+-- of the client.
+instance Encode (M.PublicKeyCredential 'M.Create) where
+  encode M.PublicKeyCredential {..} =
+    JS.PublicKeyCredential
+      { rawId = encode pkcIdentifier,
+        response = encode pkcResponse,
+        -- TODO: Extensions aren't currently supported
+        clientExtensionResults = Map.empty
+      }
+
+instance Encode (M.CollectedClientData 'M.Get) where
+  encode M.CollectedClientData {..} =
+    JS.URLEncodedBase64 . toStrict $
+      Aeson.encode
+        JS.ClientDataJSON
+          { typ = "webauthn.get",
+            challenge = Text.decodeUtf8 . Base64.encode $ M.unChallenge ccdChallenge,
+            origin = M.unOrigin ccdOrigin,
+            crossOrigin = ccdCrossOrigin
+          }
+
+instance Encode (M.AuthenticatorResponse 'M.Get) where
+  encode M.AuthenticatorAssertionResponse {..} =
+    JS.AuthenticatorAssertionResponse
+      { clientDataJSON = encode argClientData,
+        authenticatorData = JS.URLEncodedBase64 $ M.adRawData argAuthenticatorData,
+        signature = JS.URLEncodedBase64 $ M.unAssertionSignature argSignature,
+        userHandle = JS.URLEncodedBase64 . M.unUserHandle <$> argUserHandle
+      }
+
+instance Encode (M.PublicKeyCredential 'M.Get) where
+  encode M.PublicKeyCredential {..} =
+    JS.PublicKeyCredential
+      { rawId = encode pkcIdentifier,
+        response = encode pkcResponse,
+        clientExtensionResults = Map.empty
+      }
+
+-- | [(spec)](https://www.w3.org/TR/webauthn-2/#iface-authenticatorresponse)
+instance Encode (M.AuthenticatorResponse 'M.Create) where
+  encode M.AuthenticatorAttestationResponse {..} =
+    JS.AuthenticatorAttestationResponse
+      { clientDataJSON = encode arcClientData,
+        attestationObject = encode arcAttestationObject
+      }
+
+-- | [(spec)](https://www.w3.org/TR/webauthn-2/#dom-authenticatorresponse-clientdatajson)
+-- TODO: Consider if we should encode a general `CollectedClientData t`
+instance Encode (M.CollectedClientData 'M.Create) where
+  encode M.CollectedClientData {..} =
+    JS.URLEncodedBase64 . toStrict $
+      Aeson.encode
+        JS.ClientDataJSON
+          { typ = "webauthn.create",
```

Comment:
You can write a combined `SingI t => Encode (M.CollectedClientData t)`, then using
```haskell
typ = case sing @t of
  SCreate -> "webauthn.create"
  SGet -> "webauthn.get"
```